feat(secrets): document secrets management (#1385)
spinnaker/spinnaker.github.io: reference/halyard/secrets/s3-secrets.md

---
layout: single
title:  "Secrets in S3"
sidebar:
  nav: reference
---

{% include toc %}

> This document describes how to set up Spinnaker secrets in an encrypted S3 bucket. In this example, we'll be using a `mybucket` bucket in the `us-west-2` region to store GitHub credentials and a kubeconfig file. We'll be referencing the bucket by its URL `mybucket.us-west-2.amazonaws.com`.

Reviewer: ```suggestion
This document describes how to set up Spinnaker secrets in an encrypted S3 bucket. This example uses a bucket (`mybucket`) in the `us-west-2` region to store GitHub credentials and a kubeconfig file. You reference the bucket by its URL `mybucket.us-west-2.amazonaws.com`.
```

Reviewer: @KathrynLewis I know I'm killing a lot of "we"s. I think "we," "us," "let's," etc. are great in tutorials and such, but I think it's better to use 2nd person in most documentation.

Author: No problem! I agree it sounds better in 2nd person

## Authorization

> Since we're storing sensitive information, we'll protect the bucket by restricting access and [enabling encryption](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/default-bucket-encryption.html).

Reviewer: ```suggestion
Since you're storing sensitive information, you protect the bucket by restricting access and [enabling encryption](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/default-bucket-encryption.html).
```

> The important thing to remember is to run Halyard's daemon and Spinnaker services with IAM roles that allow them to read that content.

Reviewer: ```suggestion
Remember to run Halyard's daemon and Spinnaker services with IAM roles that allow them to read that content.
```

> ## Storing secrets

Reviewer: Add a blank line between these two headings.

### Storing credentials

> Let's store our GitHub credentials in `mybucket/spinnaker-secrets.yml`:

Reviewer: ```suggestion
Store your GitHub credentials in `mybucket/spinnaker-secrets.yml`:
```

```yaml
github:
  password: <PASSWORD>
  token: <TOKEN>
```

> Note: We could have chosen to store the password under different keys than `github.password` and `github.token`. We'd just need to change how to reference the secret further down.

Reviewer: ```suggestion
Note: You could chose to store the password under different keys than `github.password` and `github.token`. You'd just need to change how to reference the secret further down.
```

### Storing sensitive files

> Some of Spinnaker configuration also uses information stored as files. Let's upload the `kubeconfig` file of our Kubernetes account directly to `mybucket/mykubeconfig`.

Reviewer: ```suggestion
Some Spinnaker configuration uses information stored as files. For example, upload the `kubeconfig` file of your Kubernetes account directly to `mybucket/mykubeconfig`.
```

> 

Reviewer: Did you mean to provide an example of how to do this?

## Referencing secrets

> Now that secrets are safely stored in our bucket, we'll reference them from our config files with the following format. The S3 specific parameters (`r:<region>`, `b:<bucket>`, etc) can be in any order:

Reviewer: ```suggestion
Now that secrets are safely stored in the bucket, you reference them from your config files using the following format. The S3 specific parameters (`r:<region>`, `b:<bucket>`, etc) can be in any order:
```

```
encrypted:s3!r:<region>!b:<bucket>!f:<path to file>!k:<optional key>
```

> For example, to reference `github.password` from the file above, we'll use:

Reviewer: ```suggestion
For example, to reference `github.password` from the file above, use:
```

```

> encrypted:s3!b:mybucket!f:spinnaker-secrets.yml!r:us-west-2!k:github.password

Reviewer 2: A hal command example would be useful here too!
